components/common/src/main/java/com/hotels/styx/common/FailureHandlingTaskProcessor.java

```
@@ -0,0 +1,144 @@
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Allows various types of failure handling to be applied to a task running over a list of inputs.
+ *
+ * It can be as strict or lenient as desired - from failing the whole run immediately as soon as a single iteration fails,
+ * to just logging failures and returning successful results.
+ *
+ * The behaviour is configured by injecting lambdas when an instance of this class is built using the {@link Builder}.
+ *
+ * @param <T> input type
+ * @param <R> output type
+ */
+public class FailureHandlingTaskProcessor<T, R> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(FailureHandlingTaskProcessor.class);
+
+    private final BiConsumer<T, Exception> onEachFailure;
+    private final Consumer<Map<T, Exception>> failuresPostProcessing;
+
+    private FailureHandlingTaskProcessor(Builder<T, R> builder) {
+        this.onEachFailure = builder.onEachFailure;
+        this.failuresPostProcessing = builder.failuresPostProcessing;
+    }
+
+    /**
+     * Execute the task on a list of inputs, using the configured failure handling.
+     *
+     * @param inputs a list of inputs
+     * @param task a task to execute
+     * @return a list of outputsl
+     */
+    public List<R> process(List<T> inputs, Task<T, R> task) {
+        List<R> successes = new ArrayList<>();
+        Map<T, Exception> failures = new LinkedHashMap<>();
+
+        inputs.forEach(input -> {
+            try {
+                successes.add(task.execute(input));
+            } catch (Exception t) {
+                onEachFailure.accept(input, t);
+                failures.put(input, t);
+            }
+        });
+
+        if (!failures.isEmpty()) {
+            failuresPostProcessing.accept(failures);
+        }
+
+        return successes;
```

do you think it would make sense to change the default behaviour to 'failure' instead of just returning the number of errors?